### PR TITLE
Crypto module implementation

### PR DESCRIPTION
**Pull request by Xavrax**
feat(crypto): Add crypto module

Update the crypto module structure and add enhanced AES-CBC cryptor.

fix(crypto): Improved security of crypto implementation by increasing the cipher key entropy by a factor of two.

Improved security of crypto implementation by increasing the cipher key entropy by a factor of two.

fix: fixed missing return from failed function

Fixed missing return from failed `pbaes256_decrypt_alloc()` function

### CODE REVIEW

**Review by budgetpreneur — CHANGES_REQUESTED** [review]
Getting compile errors on windows "openssl" and "cpp" folders. Sent logs to you via slack.

File: core/pbcc_crypto_legacy.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+static int legacy_decrypt(
+        struct pubnub_cryptor_t const *algo,
+        pubnub_bymebl_t *result,
+        struct pubnub_encrypted_data to_decrypt
+) {
+    struct legacy_context *ctx = (struct legacy_context *)algo->user_data;
+
+    size_t estimated_size = estimated_dec_buffer_size(to_decrypt.data.size); // TODO: WHY!?!?!?
```
Comment (budgetpreneur) [inline core/pbcc_crypto_legacy.c]: do we need TODO comment?

File: core/pubnub_ccore_pubsub.c [hunk 2]
```diff
@@ -504,6 +542,45 @@ enum pubnub_res pbcc_publish_prep(struct pbcc_context* pb,
...
+        pubnub_bymebl_t message_block;
+        message_block.ptr = (uint8_t*)message;
+        message_block.size = strlen(message);
+
+        pubnub_bymebl_t encrypted = pb->crypto_module->encrypt(pb->crypto_module, message_block);
+
+        if (NULL == encrypted.ptr) {
+            PUBNUB_LOG_ERROR("pbcc_publish_prep(pbcc=%p) - encryption failed\n", pb);
+            free(message_block.ptr);
+            // TODO: add a specific error code for this
```
Comment (budgetpreneur) [inline core/pubnub_ccore_pubsub.c:557]: any thing to do here because of TODO?

**Review by budgetpreneur — APPROVED** [review]
ok

File: core/pbcc_crypto_legacy.c (hunk 1 above)
Comment (Xavrax) [reply core/pbcc_crypto_legacy.c]: No, Ill remove that.

File: core/pubnub_ccore_pubsub.c (hunk 2 above)
Comment (Xavrax) [reply core/pubnub_ccore_pubsub.c:557]: I'll remove that - I found that we just return this error in case of encryption.

**Comment by Xavrax** [comment]
@pubnub-release-bot release

**Pull request merged**